### ISSUE TITLE
Commands auto-completion support

### ISSUE THREAD

**Issue by bochaco** [maidsafe/safe-cli#73]
Auto completion even coming from SAFE data (for ref: https://www.joshmcguigan.com/blog/shell-completions-pure-rust/)

**Comment by bochaco** [comment]
A first PoC was implemented by community member @draw here: https://safenetforum.org/t/new-release-vault-phase-1-real-vault/29712/145

**Comment by dan-da** [comment]
I've started working on this.  I added a --dump-completions=<shell> flag to safe-cli which dumps shell script that can be sourced by bash, zsh, etc.  This functionality is built into clap, so it automatically uses the values from structopt.  Very elegant.

Activating the completions for bash looks like:

```
safe --dump-completions=bash > /tmp/out && source /tmp/out
```

After that commands, sub-commands and flags auto complete.  slick.  examples:

```
$ safe <tab>
auth                --dry-run           -h                  keypair             -o                  -V
cat                 --dump-completions  help                keys                --output            --version
container           --endpoint          --help              -n                  safe-id             wallet
dog                 files               --json              nrs                 update              --xorurl
```

```
$ safe files 
add                 --endpoint          --help              -o                  sync                --xorurl
--dry-run           -h                  --json              --output            -V                  
--dump-completions  help                -n                  put                 --version
```

```
$ safe cat 
--dry-run           -h                  <location>          --output            --xorurl            
--dump-completions  --help              -n                  -V                  
--endpoint          --json              -o                  --version
```

Ideally at install time,  --dump-completions would be run once for each possible shell and then installed into proper location for user (or system-wide).

Also --dump-completions might be something we only want to build into dev/debug releases.   thoughts advice on that?

**Comment by bochaco** [comment]
That looks really neat, and I agree it's really good if it's generated from the structopt so it can be done with any version of the CLI. I'm wondering if we should have a flag e.g. `--auto-completion` which does the dump and source? Do you know if this will work on Windows too? or if we can implement the Windows equivalent under the same flag?

**Comment by wardbr** [comment]
I had a very quick look in the past to clap. For autocompletion of the static start arguments: that seems rather easy and elegant. But it becomes more complicated when you implement dynamic stuff, like 'safe cat safe://foo/<tab>', which I did get working. Especially the ':'-character gives troubles, because that is default a Field Seperator in the bash autocompletion mechansim. So you have to de/re activate ':' as field seperator on the correct moment. Probably something similar if you want also to support autocompletion for spaces in the 'safe://...' url. That last one, I haven't implemented, but I know you could certainly solve it with escape with '\' like "safe cat safe://foo/bar\ <tab>".
You can implement this stuff in Rust itself and my (and other) autcompletion scripts in Bash can help with that to know what to do when.

**Comment by dan-da** [comment]
@bochaco hmm, auto-completion flag to dump and source is a bit tricky because then we have to figure out what shell user is using and exec proper command(s) for that shell.

Yes, it should work on windows with powershell. I haven't tried it myself, but I see powershell is supported and windows support is specifically mentioned in the changelog a few times..   See the full list of [shells](https://github.com/clap-rs/clap_generate/tree/master/src/shells).  To dump powershell version would be:

```
safe --dump-completions=powershell
```

I was musing about making it available only under dev/debug because it is not really relevant to core mission/features of safe-cli (or whichever executable).  Rather it is something that could potentially be used only during the build process, and then have the generated completion files, eg bash_completions.bash distributed and installed.   That said, for now easiest thing is just to leave it in there always.  I was just throwing it out as a possibility.

**Comment by dan-da** [comment]
@wardbr yeah I'm trying not to let perfect be the enemy of the good here.

What you are talking about with auto-completing dynamic stuff like URLs sounds cool and should be possible inside the safe-cli shell itself, but would be quite tricky or maybe even impossible in many/most standard shells when invoking safe-cli.   One would have to either implement all the logic in the (eg) bash code itself, or else have bash code exec the rust process for each char typed.

What I have working now is already quite useful and was quick/easy to do with just a few lines of code.  I think it will satisfy most user's expectations.   It could always be improved upon in the future, but yeah I think it is important to separate in our minds completion at the bash/invocation level vs completion in the internal shell.  My code is aimed only at the invocation level.  It might be possible to leverage clap_generate for the internal shell, but I'm not certain yet that even makes sense.

**Comment by bochaco** [comment]
ah right I forgot you need to specify that in the source command, fair enough.

> Rather it is something that could potentially be used only during the build process, and then have the generated completion files, eg bash_completions.bash distributed and installed

Right, I think as you also mentioned it's not bad to allow users to simply generate it rather than download it, so we don't really need to publish it but just explain users how they can dump it and use it. This makes our release simpler, just one binary

**Comment by wardbr** [comment]
@dan-da To be clear: https://safenetforum.org/t/new-release-vault-phase-1-real-vault/29712/145 already supports the cool dynamic auto-completing URL + only show the start arg when applicable. This is for bash only and if arguments are changed in the Rust source code, you have to manually update .safe_completion.sh at the moment. But that is not that much work, I think.
I also think it has not much added value to implement  this auto-completion URL stuff in Rust (for Bash/Zsh command line).

**Comment by bochaco** [comment]
wrt to safe-URL auto completion, it will be nice, but as I think @wardbr says it doesn't have to be done now. I also imagined if it was possible to create a driver that can be installed on the OS which makes everything to work with safe URLs, so if I wan to launch a program stored on SAFE I simply can do `$ safe:///myprograms/solitary`, if so, then URL auto-completion could also work through such a driver if it fetches the info needed by the OS...??...

**Comment by wardbr** [comment]
Concerning (safe-URL) auto completion: nothing has to be done now in my opinion, so I don't want to waste too much of your time ;)
Let me summarize what the main thing I wanted to say:
- Not much added value: something like https://www.joshmcguigan.com/blog/shell-completions-pure-rust/
- Advantages automatic generation of Bash autocompletion with 'safe --dump-completions':
  + binary enough for getting autocompletion
  + Developers don't have to update autocompletion script seperately.
- Advantage of  https://safenetforum.org/t/new-release-vault-phase-1-real-vault/29712/145
  + Has already support for safe-URL auto completion.

**Comment by dan-da** [comment]
@wardbr thx for sharing link to Josh McGuigan's blog.  That's pretty cool that there's a nice mechanism already for generating dynamic completions for things like URLs within an arbitrary executable. I found he also has a rust crate [shell_completion](https://github.com/JoshMcguigan/shell_completion) available, though it doesn't *yet* integrate with structopt or provide any other shells/platforms than bash.

I think that for now, I will do a bit of cleanup and submit what I have as a PR for consideration/review and we can continue discussion there.
